### PR TITLE
SONAR-6381 add java WS to install a plugin

### PR DESCRIPTION
**Pull request by sns-seb**
just the WS glue to the underlying components.
note : only the latest version of the plugin can be installed

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/plugins/ws/InstallPluginsWsAction.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+  @Override
+  public void define(WebService.NewController controller) {
+    WebService.NewAction action = controller.createAction("install")
+      .setPost(true)
+      .setDescription("Installs the latest version of a plugin specified by its key.")
+      .setHandler(this);
+
+    action.createParam(PARAM_KEY).setRequired(true)
+      .setDescription("The key identifying the plugin to install");
```
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/plugins/ws/InstallPluginsWsAction.java]: Minor feedback: can you provide an example value?

File: server/sonar-server/src/main/java/org/sonar/server/plugins/ws/InstallPluginsWsAction.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    String key = request.mandatoryParam(PARAM_KEY);
+    PluginUpdate pluginUpdate = findAvailablePluginByKey(key);
+    pluginDownloader.download(key, pluginUpdate.getRelease().getVersion());
+    response.noContent();
+  }
+
+  @Nonnull
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/plugins/ws/InstallPluginsWsAction.java]: No need for @Nonnull annotation on method returns as this is the default behavior. @CheckForNull should be used when result can be null.

**Comment by simonbrandhof** [comment]
LGTM

File: server/sonar-server/src/main/java/org/sonar/server/plugins/ws/InstallPluginsWsAction.java (hunk 1 above)
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/plugins/ws/InstallPluginsWsAction.java]: This WS does not return any JSON response. We could think about documenting such behavior, but how ?

File: server/sonar-server/src/main/java/org/sonar/server/plugins/ws/InstallPluginsWsAction.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    PluginDownloader pluginDownloader) {
+    this.updateCenterFactory = updateCenterFactory;
+    this.pluginDownloader = pluginDownloader;
+  }
+
+  @Override
+  public void define(WebService.NewController controller) {
+    WebService.NewAction action = controller.createAction("install")
+      .setPost(true)
+      .setDescription("Installs the latest version of a plugin specified by its key.")
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/plugins/ws/InstallPluginsWsAction.java]: You should maybe mention that the plugin is downloaded from update center.

File: server/sonar-server/src/main/java/org/sonar/server/plugins/ws/InstallPluginsWsAction.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      .setHandler(this);
+
+    action.createParam(PARAM_KEY).setRequired(true)
+      .setDescription("The key identifying the plugin to install");
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/plugins/ws/InstallPluginsWsAction.java]: Security - check that user is a system administrator.

**Comment by simonbrandhof** [comment]
Removed my previous LGTM comment - see blocker comment about permission

**Comment by sns-seb** [comment]
PR update with blocking comment taken into account and rebase on master

File: server/sonar-server/src/main/java/org/sonar/server/plugins/ws/InstallPluginsWsAction.java (hunk 2 above)
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/plugins/ws/InstallPluginsWsAction.java]: +1

**Comment by simonbrandhof** [comment]
Added comment on description + LGTM

**Pull request merged**